### DISCUSSION TITLE
Set Filter on new Data

### DISCUSSION THREAD

**Discussion by vaclav-domin** [dotnet/DataGridExtensions#60] [Q&A] [answered]
I need apply filter after loading data
Now:
Load data (Obsarvable collection)
Create CollectionViewSource
Apply saved filter to columns (column.SetFilter(savedfilterforcolumn))

Filters on columns are visible, but filters don't apply.
Missing solution or missing step?

Thanks

**Comment by tom-englert** [comment]
Can you provide a small sample project to reproduce this?

**Comment by tom-englert** [comment, marked as answer]
Did you do it the same way as in the sample?
https://github.com/dotnet/DataGridExtensions/blob/master/src/DataGridExtensionsSample/Views/BasicView.xaml.cs#L21-L27

**Reply by vaclav-domin** [reply]
Yes, this is solution!
Thanks

**Reply by vaclav-domin** [reply]
Oh now.
Sample code

        public virtual void AfterLoadData()
        {
            if (Data == null) return;
            if (ViewSource != null) ViewSource.CurrentChanged -= ViewSource_CurrentChanged;

            ViewSource = (ListCollectionView)CollectionViewSource.GetDefaultView(Data);
            //obnovit seskupení
            var gc = GridColumns.Where(x => x.SortDirection != null).ToList();
            ViewSource.CurrentChanged += ViewSource_CurrentChanged;
            foreach (var c in gc)
            {
                ViewSource.SortDescriptions.Add(
                        new SortDescription(c.SortMemberPath, c.SortDirection.Value)
                    );
            };
            GroupInfoList.Where(x => x.IsGrouping).OrderBy(x => x.GroupingLevel).ToList()
                .ForEach(x => { ViewSource.GroupDescriptions.Add(new PropertyGroupDescription(x.ColumnName)); });
            
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("Data"));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(SavedSelectedViewName)));
            
            //SetGridFocus();
            System.Windows.Threading.Dispatcher.CurrentDispatcher.Invoke( new Action(() => { 
                    RestoreColumns(); 
                    ViewSource.Refresh();
            }), System.Windows.Threading.DispatcherPriority.ContextIdle);
            ViewSource.Refresh();
        }




        public virtual void RestoreColumns()
        {
                try
                {
                    List<ColumnInfo> columnInfoList = DataGridViewSettings.Default.ColumnSettings;
                    if (columnInfoList != null)
                    {
                        var sorted = columnInfoList.OrderBy(x => x.DisplayIndex);
                        foreach (var itm in sorted)
                        {
                            GridColumns[itm.ColumnIndex].DisplayIndex = itm.DisplayIndex;
                            GridColumns[itm.ColumnIndex].Visibility = itm.Visible ? Visibility.Visible : Visibility.Hidden;
                            GridColumns[itm.ColumnIndex].Width = new DataGridLength(itm.Width, itm.UnitType);
                            GridColumns[itm.ColumnIndex].SortDirection = itm.SortDirection;
                            GridColumns[itm.ColumnIndex].SetFilter(itm.ColumnFilter);
                        }
                    }
                }
                catch
                {
                }
                finally 
                {
                }
            }

        }
